src/prefect/agent.py

```diff
@@ -177,6 +189,85 @@ async def get_and_submit_flow_runs(self) -> List[FlowRun]:
 
         return submittable_runs
 
+    async def check_for_cancelled_flow_runs(self):
+        if not self.started:
+            raise RuntimeError("Agent is not started. Use `async with OrionAgent()...`")
+
+        self.logger.debug("Checking for cancelled flow runs...")
+
+        work_queue_names = set()
+        async for work_queue in self.get_work_queues():
+            work_queue_names.add(work_queue.name)
+
+        cancelling_flow_runs = await self.client.read_flow_runs(
+            flow_run_filter=FlowRunFilter(
+                state=FlowRunFilterState(
+                    name=FlowRunFilterStateName(any_=["Cancelling"])
+                ),
+                work_queue_name=FlowRunFilterWorkQueueName(any_=list(work_queue_names)),
+            ),
+        )
+
+        if cancelling_flow_runs:
+            self.logger.info(
+                f"Found {len(cancelling_flow_runs)} flow runs awaiting cancellation."
+            )
+
+        async with anyio.create_task_group() as tg:
+            for flow_run in cancelling_flow_runs:
+
+                # Avoid duplicate cancellation calls
+                if flow_run.id in self.cancelling_flow_run_ids:
+                    continue
+                else:
+                    self.cancelling_flow_run_ids.add(flow_run.id)
+
+                tg.start_soon(self.cancel_run, flow_run)
+
+        return cancelling_flow_runs
+
+    async def cancel_run(self, flow_run: FlowRun) -> None:
+        """
+        Cancel a flow run by killing its infrastructure
+        """
+        try:
+            infrastructure = await self.get_infrastructure(flow_run)
+        except Exception:
+            self.logger.exception(
+                f"Failed to get infrastructure for flow run '{flow_run.id}'. "
+                "Flow run cannot be cancelled."
+            )
+            # Note: We leave this flow run in the cancelling set because it cannot be
+            #       cancelled and this will prevent additional attempts.
+            return
+
+        async def mark_flow_run_as_cancelled():
+            state = flow_run.state.copy(update={"name": "Cancelled"})
+            await self.client.set_flow_run_state(flow_run.id, state, force=True)
+            self.cancelling_flow_run_ids.remove(flow_run.id)
+
+        self.logger.info(
+            f"Killing {infrastructure.type} {flow_run.infrastructure_pid} for flow run '{flow_run.id}'..."
+        )
+        try:
+            await infrastructure.kill(flow_run.infrastructure_pid)
```

Comment:
Should we handle the case that `infrastructure_pid` is null?

Reply:
Yeah let's handle that and do a `hasattr` check for `kill`.

Reply:
On second thought, we should maybe handle this during the proposal of a cancelled state? We can add a `kill-infrastructure` capability to infrastructure blocks that support killing and block cancellation when it is not supported?

Reply:
I added these in https://github.com/PrefectHQ/prefect/pull/7671/commits/52f5565392b3d33257a373399a00a18343c10360 and we mark the run as cancelled so we don't pick it up forever. Kind of an awkward situation.